### ISSUE TITLE
KotlinGenerator produces wrong code with kotlinNotNullPojoAttributes when multiple references to an embeddable have different nullability

### ISSUE THREAD

**Issue by F43nd1r** [jOOQ/jOOQ#14991]
### Expected behavior

Note: This is a follow-up to https://github.com/jOOQ/jOOQ/issues/14853

Generated kotlin POJOs compile when using embeddables, e.g.

```kotlin
data class Bug(
    override val latestVersionCode: Int? = null,
    override val latestVersionFlavor: String? = null,
): IBug {
    override val latestVersionKey: IVersionKey?
        get(): IVersionKey? = if(latestVersionCode != null && latestVersionFlavor != null) VersionKey(
            latestVersionCode,
            latestVersionFlavor
        ) else null
}

data class VersionKey(
    override val code: Int,
    override val flavor: String
): IVersionKey
```

or
```kotlin
data class Bug(
    override val latestVersionCode: Int? = null,
    override val latestVersionFlavor: String? = null,
): IBug {
    override val latestVersionKey: IVersionKey
        get(): IVersionKey = VersionKey(
            latestVersionCode,
            latestVersionFlavor
        )
}

data class VersionKey(
    override val code: Int? = null,
    override val flavor: String? = null
): IVersionKey
```

First version would be preferred by me. Maybe configurable?

### Actual behavior

```kotlin
data class Bug(
    override val latestVersionCode: Int? = null,
    override val latestVersionFlavor: String? = null,
): IBug {
    override val latestVersionKey: IVersionKey
        get(): IVersionKey = VersionKey(
            latestVersionCode, //compilation fails here with "Type mismatch. Required: Int Found: Int?"
            latestVersionFlavor
        )
}

data class VersionKey(
    override val code: Int,
    override val flavor: String
): IVersionKey
```

### Steps to reproduce the problem

Used configuration:
```kotlin
isImmutableInterfaces = true
isKotlinNotNullInterfaceAttributes = true

isKotlinNotNullRecordAttributes = true

isPojos = true
isPojosAsKotlinDataClasses = true
isKotlinNotNullPojoAttributes = true
isImmutablePojos = true
```
```kotlin
EmbeddableDefinitionType().apply {
    name = "VERSION_KEY"
    referencingName = "LATEST_VERSION_KEY"
    tables = "BUG"
    fields = listOf(
        EmbeddableField().withName("CODE").withExpression("LATEST_VERSION_CODE"),
        EmbeddableField().withName("FLAVOR").withExpression("LATEST_VERSION_FLAVOR"),
    )
}
```

`LATEST_VERSION_CODE` and `LATEST_VERSION_FLAVOR` are nullable in the schema.

### jOOQ Version

3.18.3

### Database product and version

mysql 8.0.31

### Java Version

openjdk version "17.0.6" 2023-01-17

### OS Version

Arch Linux

### JDBC driver name and version (include name if unofficial driver)

 com.mysql:mysql-connector-j:8.0.32

**Comment by lukaseder** [comment]
Thanks a lot for your report. I'll try reproducing this soon.

**Comment by lukaseder** [comment]
The integration test I added for https://github.com/jOOQ/jOOQ/issues/14853 and https://github.com/jOOQ/jOOQ/issues/14855 already contains all the flags you've documented:

```xml
<pojos>true</pojos>                                                      
<pojosAsKotlinDataClasses>true</pojosAsKotlinDataClasses>                
<records>true</records>                                                  
<interfaces>true</interfaces>                                            
<immutablePojos>true</immutablePojos>                                    
<immutableInterfaces>true</immutableInterfaces>                          
<kotlinNotNullPojoAttributes>true</kotlinNotNullPojoAttributes>          
<kotlinNotNullRecordAttributes>true</kotlinNotNullRecordAttributes>      
<kotlinNotNullInterfaceAttributes>true</kotlinNotNullInterfaceAttributes>
```

The test case doesn't reproduce your problem. Can you please provide your `CREATE TABLE` statement and embeddable configuration? (Ideally, a complete self contained reproducer would be even better, but perhaps `CREATE TABLE` and embeddable configuration suffices to reproduce the problem)

**Comment by F43nd1r** [comment]
While certainly not minimal, my project is open source and I've pushed a commit with the failing generation [here](https://github.com/F43nd1r/Acrarium/tree/jooq-kotlin).

Table DDL:
```sql
CREATE TABLE `bug` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `title` longtext NOT NULL,
  `app_id` int(11) NOT NULL,
  `report_count` int(11) NOT NULL DEFAULT 0,
  `latest_report` datetime DEFAULT NULL,
  `solved_version_code` int(11) DEFAULT NULL,
  `solved_version_flavor` varchar(255) DEFAULT NULL,
  `latest_version_code` int(11) DEFAULT NULL,
  `latest_version_flavor` varchar(255) DEFAULT NULL,
  `affected_installations` int(11) NOT NULL DEFAULT 0,
  PRIMARY KEY (`id`),
  KEY `FK_bug_solved_version` (`app_id`,`solved_version_code`,`solved_version_flavor`),
  KEY `FK_bug_latest_version` (`app_id`,`latest_version_code`,`latest_version_flavor`),
  CONSTRAINT `FK_bug_app` FOREIGN KEY (`app_id`) REFERENCES `app` (`id`) ON DELETE CASCADE,
  CONSTRAINT `FK_bug_latest_version` FOREIGN KEY (`app_id`, `latest_version_code`, `latest_version_flavor`) REFERENCES `version` (`app_id`, `code`, `flavor`),
  CONSTRAINT `FK_bug_solved_version` FOREIGN KEY (`app_id`, `solved_version_code`, `solved_version_flavor`) REFERENCES `version` (`app_id`, `code`, `flavor`)
) ENGINE=InnoDB AUTO_INCREMENT=23906 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```

Full generator config:
```kotlin
name = "org.jooq.codegen.KotlinGenerator"
database.apply {
    name = "com.faendir.jooq.MySqlLiquibaseDatabase"
    includes = ".*"
    excludes = "DATABASECHANGELOG|DATABASECHANGELOGLOCK"
    inputSchema = "acrarium"
    properties = listOf(
        Property().withKey("databaseName").withValue(inputSchema),
        Property().withKey("scripts").withValue(changelogPath),
    )
    forcedTypes = listOf(
        ForcedType().apply {
            userType = "com.faendir.acra.persistence.app.AppId"
            converter = "com.faendir.acra.persistence.app.AppIdConverter"
            includeExpression = ".*\\.app_id|app.id"
        },
        ForcedType().apply {
            userType = "com.faendir.acra.persistence.bug.BugId"
            converter = "com.faendir.acra.persistence.bug.BugIdConverter"
            includeExpression = ".*\\.bug_id|bug.id"
        },
        ForcedType().apply {
            userType = "java.time.Instant"
            converter = "com.faendir.acra.persistence.jooq.InstantConverter"
            includeTypes = "DATETIME"
        },
    )
    embeddables = listOf(
        EmbeddableDefinitionType().apply {
            name = "VERSION_KEY"
            tables = "REPORT"
            fields = listOf(
                EmbeddableField().withName("CODE").withExpression("VERSION_CODE"),
                EmbeddableField().withName("FLAVOR").withExpression("VERSION_FLAVOR"),
            )
        },
        EmbeddableDefinitionType().apply {
            name = "VERSION_KEY"
            referencingName = "LATEST_VERSION_KEY"
            tables = "BUG"
            fields = listOf(
                EmbeddableField().withName("CODE").withExpression("LATEST_VERSION_CODE"),
                EmbeddableField().withName("FLAVOR").withExpression("LATEST_VERSION_FLAVOR"),
            )
        },
        EmbeddableDefinitionType().apply {
            name = "VERSION_KEY"
            referencingName = "SOLVED_VERSION_KEY"
            tables = "BUG"
            fields = listOf(
                EmbeddableField().withName("CODE").withExpression("SOLVED_VERSION_CODE"),
                EmbeddableField().withName("FLAVOR").withExpression("SOLVED_VERSION_FLAVOR"),
            )
        },
    )
}
target.apply {
    packageName = "com.faendir.acra.jooq.generated"
    directory = "$buildDir/generated/source/jooq/main"
}
generate.apply {
    isImmutableInterfaces = true
    isKotlinNotNullInterfaceAttributes = true

    isKotlinNotNullRecordAttributes = true

    isPojos = true
    isPojosAsKotlinDataClasses = true
    isKotlinNotNullPojoAttributes = true
    isImmutablePojos = true
}
```

**Comment by lukaseder** [comment]
Looking into this now. What does `com.faendir.jooq.MySqlLiquibaseDatabase` do?

**Comment by lukaseder** [comment]
Another attempt at reproducing this failed. I had to make tons of assumptions from your bits of information, but apparently, that's not sufficient. This:

```xml
<execution>
    <id>generate-thebug</id>
    <phase>generate-test-sources</phase>
    <goals>
        <goal>generate</goal>
    </goals>
    <configuration>
        <generator>
            <database>
                <name>org.jooq.meta.extensions.ddl.DDLDatabase</name>
                <properties>
                    <property>
                        <key>sql</key>
                        <value>
                            CREATE TABLE `report` (
                              `app_id` int NOT NULL,
                              `code` int NOT NULL,
                              `flavor` int NOT NULL,
                              PRIMARY KEY (`app_id`, `code`, `flavor`)
                            );

                            CREATE TABLE `version` (
                            `app_id` int NOT NULL,
                            `code` int NOT NULL,
                            `flavor` int NOT NULL,
                            PRIMARY KEY (`app_id`, `code`, `flavor`)
                            );
                            CREATE TABLE `bug` (
                            `id` int(11) NOT NULL AUTO_INCREMENT,
                            `title` longtext NOT NULL,
                            `app_id` int(11) NOT NULL,
                            `report_count` int(11) NOT NULL DEFAULT 0,
                            `latest_report` datetime DEFAULT NULL,
                            `solved_version_code` int(11) DEFAULT NULL,
                            `solved_version_flavor` varchar(255) DEFAULT NULL,
                            `latest_version_code` int(11) DEFAULT NULL,
                            `latest_version_flavor` varchar(255) DEFAULT NULL,
                            `affected_installations` int(11) NOT NULL DEFAULT 0,
                            PRIMARY KEY (`id`),
                            KEY `FK_bug_solved_version` (`app_id`,`solved_version_code`,`solved_version_flavor`),
                            KEY `FK_bug_latest_version` (`app_id`,`latest_version_code`,`latest_version_flavor`),
                            CONSTRAINT `FK_bug_latest_version` FOREIGN KEY (`app_id`, `latest_version_code`, `latest_version_flavor`) REFERENCES `version` (`app_id`, `code`, `flavor`),
                            CONSTRAINT `FK_bug_solved_version` FOREIGN KEY (`app_id`, `solved_version_code`, `solved_version_flavor`) REFERENCES `version` (`app_id`, `code`, `flavor`)
                            )
                        </value>
                    </property>
                </properties>
                <forcedTypes>
                    <forcedType>
                        <userType>com.faendir.acra.persistence.app.AppId</userType>
                        <converter>com.faendir.acra.persistence.app.AppIdConverter</converter>
                        <includeExpression>.*\\.app_id|app.id</includeExpression>
                    </forcedType>
                    <forcedType>
                        <userType>com.faendir.acra.persistence.bug.BugId</userType>
                        <converter>com.faendir.acra.persistence.bug.BugIdConverter</converter>
                        <includeExpression>.*\\.bug_id|bug.id</includeExpression>
                    </forcedType>
                    <forcedType>
                        <userType>java.time.Instant</userType>
                        <converter>com.faendir.acra.persistence.jooq.InstantConverter</converter>
                        <includeTypes>DATETIME</includeTypes>
                    </forcedType>
                </forcedTypes>
                <embeddables>
                    <embeddable>
                        <name>version_key</name>
                        <tables>report</tables>
                        <fields>
                            <field><name>code</name><expression>version_code</expression></field>
                            <field><name>flavor</name><expression>version_flavor</expression></field>
                        </fields>
                    </embeddable>
                    <embeddable>
                        <name>version_key</name>
                        <referencingName>LATEST_VERSION_KEY</referencingName>
                        <tables>bug</tables>
                        <fields>
                            <field><name>code</name><expression>latest_version_code</expression></field>
                            <field><name>flavor</name><expression>latest_version_flavor</expression></field>
                        </fields>
                    </embeddable>
                    <embeddable>
                        <name>version_key</name>
                        <referencingName>SOLVED_VERSION_KEY</referencingName>
                        <tables>bug</tables>
                        <fields>
                            <field><name>code</name><expression>solved_version_code</expression></field>
                            <field><name>flavor</name><expression>solved_version_flavor</expression></field>
                        </fields>
                    </embeddable>
                </embeddables>
            </database>
            <generate>
                <records>true</records>
                <immutableInterfaces>true</immutableInterfaces>
                <kotlinNotNullInterfaceAttributes>true</kotlinNotNullInterfaceAttributes>
                <kotlinNotNullRecordAttributes>true</kotlinNotNullRecordAttributes>
                <pojos>true</pojos>
                <pojosAsKotlinDataClasses>true</pojosAsKotlinDataClasses>
                <kotlinNotNullPojoAttributes>true</kotlinNotNullPojoAttributes>
                <immutablePojos>true</immutablePojos>
            </generate>
            <target>
                <packageName>org.jooq.codegen.test.thebug.db</packageName>
            </target>
        </generator>
    </configuration>
</execution>
```

Generates compilation errors due to lacking converters (not sure if those are relevant), but I don't get the error you're referring to.

I'll really need a complete, self-contained reproducer to track down the problem. We have a template for this, it includes a kotlin codegen sample configuration:
https://github.com/jOOQ/jOOQ-mcve

**Comment by F43nd1r** [comment]
I'll see what I can do regarding minimizing reproducers. Thanks for the template, should help.

**Comment by lukaseder** [comment]
OK, I get it. There's probably a mismatch between nullabilities of declaring columns for the embeddable and referencing columns. This seems to reproduce it:

```xml
<configuration>
    <generator>
        <database>
            <name>org.jooq.meta.extensions.ddl.DDLDatabase</name>
            <properties>
                <property>
                    <key>sql</key>
                    <value>
                        CREATE TABLE a (
                          x int DEFAULT NULL,
                          y varchar(255) DEFAULT NULL
                        );
                        CREATE TABLE b (
                          x int NOT NULL,
                          y varchar(255) NOT NULL
                        );
                    </value>
                </property>
            </properties>
            <embeddables>
                <embeddable>
                    <name>e</name>
                    <tables>a</tables>
                    <fields>
                        <field><name>x</name><expression>x</expression></field>
                        <field><name>y</name><expression>y</expression></field>
                    </fields>
                </embeddable>
                <embeddable>
                    <name>e</name>
                    <referencingName>e</referencingName>
                    <tables>b</tables>
                    <fields>
                        <field><name>x</name><expression>x</expression></field>
                        <field><name>y</name><expression>y</expression></field>
                    </fields>
                </embeddable>
            </embeddables>
        </database>
        <generate>
            <records>true</records>
            <immutableInterfaces>true</immutableInterfaces>
            <kotlinNotNullInterfaceAttributes>true</kotlinNotNullInterfaceAttributes>
            <kotlinNotNullRecordAttributes>true</kotlinNotNullRecordAttributes>
            <pojos>true</pojos>
            <pojosAsKotlinDataClasses>true</pojosAsKotlinDataClasses>
            <kotlinNotNullPojoAttributes>true</kotlinNotNullPojoAttributes>
            <immutablePojos>true</immutablePojos>
        </generate>
        <target>
            <packageName>org.jooq.codegen.test.thebug.db</packageName>
        </target>
    </generator>
</configuration>
```

The generated embeddable is this:

```kotlin
internal data class E(
    override val x: Int,
    override val y: String
): IE { ... }
```

And it is referenced like this:

```kotlin
// Nullability doesn't match
internal data class A(
    override val x: Int? = null,
    override val y: String? = null
): IA {
    override val e: IE
        get(): IE = E(
            x, // compilation error
            y
        )
    ...
}

// Nullability matches
internal data class B(
    override val x: Int,
    override val y: String
): IB {
    override val e: IE
        get(): IE = E(
            x,
            y
        )
    ...
}
```

Does that look like what you're really doing?

**Comment by lukaseder** [comment]
Hmm, the assumption underlying the embeddable code generation design is that all references to the same embeddable consist of columns of compatible type. So far, only the type itself was relevant to this, not the constraints on the type, but that's obviously obsolete.

I suspect that if any reference to an embeddable column is nullable, then the embeddable column must be nullable as well, irrespective of the order of embeddable definitions.

I'll have to think about this a bit more. It certainly isn't an edge case. The embeddable covers a unique constraint (not null), but references to it may well be nullable.

**Comment by F43nd1r** [comment]
Oh, nice find! Yes every version has a code and a flavor, but not every bug has a solved version (some bugs are unsolved).

**Comment by lukaseder** [comment]
OK, so this test should cover all cases, irrespective of ordering of tables / configurations:

```sql
create table embeddables_nullability_conflict_1 (
  id int primary key not null,
  n_n text,
  n_nn text,
  nn_n text not null,
  nn_nn text not null,
  c text
);


create table embeddables_nullability_conflict_2 (
  id int primary key not null,
  n_n text,
  n_nn text not null,
  nn_n text,
  nn_nn text not null,
  c text
);
```

And then:

```xml
<embeddable>
    <name>E_CONFLICT</name>
    <tables>EMBEDDABLES_NULLABILITY_CONFLICT_(1|2)</tables>
    <fields>
        <field><expression>N_N</expression></field>
        <field><expression>N_NN</expression></field>
        <field><expression>NN_N</expression></field>
        <field><expression>NN_NN</expression></field>
    </fields>
</embeddable>
```

The generated `data class EmbeddablesNullabilityConflict_1` now has a compilation error here:

```kotlin
    override var eConflict: IEConflict
        set(value): Unit {
            nN = value.nN
            nNn = value.nNn
            nnN = value.nnN // Error here
            nnNn = value.nnNn
        }
        get(): IEConflict = EConflict(
            nN,
            nNn, // Error here
            nnN,
            nnNn
        )
```

This is because the embeddable `data class EConflict` looks like this:

```kotlin
internal data class EConflict(
    override var nN: String? = null,
    override var nNn: String,
    override var nnN: String? = null,
    override var nnNn: String
): IEConflict { .. }
```

I.e. it was generated from the second table (the last embeddable match seems to override the first). But it wouldn't matter the other way round.

You only got one of these errors (on the getter) due to the generation of immutable objects, but the same type of error will also appear on a setter.

I still think that the read case is more important. The shared embeddable must be able to handle `NULL` values originating from *any* referencing table. So, in this case, the embeddable should be generated as follows:

```kotlin
internal data class EConflict(
    override var nN: String? = null,
    override var nNn: String? = null,
    override var nnN: String? = null,
    override var nnNn: String
): IEConflict { .. }
```

I guess we cannot do much on the setter other than handle the compilation error. Either with `!!`, or maybe better, with an explicit error message hinting at the problem:

```kotlin
nNn = value.nNn ?: throw NullPointerException("Shared embeddable allows NULL value, but this table does not")
```

**Comment by lukaseder** [comment]
> (the last embeddable match seems to override the first)

Ah yes, I remember this caveat now. The two `EmbeddableDefinition` are distinct instances, because they have a distinct qualified name (`table` + `referencingName`). This leads to two classes being produced with the same qualified name, where later embeddables effectively overwrite existing classes.

I vaguely remember this being an issue elsewhere on this issue tracker, but can't find it now. In any case, this should be fixed, of course. Each class should be generated only once.

**Comment by lukaseder** [comment]
There's a separate discussion that may be worth having. Is an embeddable value consisting *only* of `null`s the same thing as that value being `null`? I.e. in your case, does it ever make sense to create a `VersionKey(null, null)`, or should the `VersionKey` itself be nullable wherever it is optional? I tend to think that the latter is more desirable in the event where *all* the references to `VersionKey` columns are nullable, but the declaration isn't.

Though, that can't be fixed easily without introducing backwards incompatibilities. I've created a separate issue for this: https://github.com/jOOQ/jOOQ/issues/15086

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.19.0 and 3.18.5 (https://github.com/jOOQ/jOOQ/issues/15087).

**Comment by F43nd1r** [comment]
> There's a separate discussion that may be worth having. Is an embeddable value consisting _only_ of `null`s the same thing as that value being `null`? I.e. in your case, does it ever make sense to create a `VersionKey(null, null)`, or should the `VersionKey` itself be nullable wherever it is optional? I tend to think that the latter is more desirable in the event where _all_ the references to `VersionKey` columns are nullable, but the declaration isn't.
> 
> Though, that can't be fixed easily without introducing backwards incompatibilities. I've created a separate issue for this: #15086

No, a version key for me is always either fully present or fully absent. I would like to see #15086 implemented as it would save me some mapping logic.
